srml/support/procedural/src/construct_runtime/mod.rs

```diff
@@ -0,0 +1,395 @@
+
+	// Assert we have system module declared
+	let system_module = match find_system_module(modules.iter()) {
+		Some(sm) => sm,
+		None => {
+			return syn::Error::new(
+				modules_token.span,
+				"`System` module declaration is missing. \
+				 Please add this line: `System: system::{Module, Call, Storage, Config, Event},`",
+			)
+			.to_compile_error()
+			.into()
+		}
+	};
+
+	let hidden_crate_name = "construct_runtime";
+	let scrate = generate_crate_access(&hidden_crate_name, "srml-support");
+	let scrate_decl = generate_hidden_includes(&hidden_crate_name, "srml-support");
+	let outer_event = try_tok!(decl_outer_event_or_origin(
+		&name,
+		modules.iter().filter(|module| module.name != "System"),
+		&system_module,
+		&scrate,
+		DeclOuterKind::Event,
+	));
+	let outer_origin = try_tok!(decl_outer_event_or_origin(
+		&name,
+		modules.iter().filter(|module| module.name != "System"),
+		&system_module,
+		&scrate,
+		DeclOuterKind::Origin,
+	));
+	let all_modules = decl_all_modules(&name, modules.iter().filter(|module| module.name != "System"));
+
+	let dispatch = decl_outer_dispatch(&name, modules.iter(), &scrate);
+	let metadata = decl_runtime_metadata(&name, modules.iter(), &scrate);
+	let outer_config = decl_outer_config(&name, modules.iter(), &scrate);
+	let inherent = decl_outer_inherent(&block, &unchecked_extrinsic, modules.iter(), &scrate);
+	let validate_unsigned = decl_validate_unsigned(&name, modules.iter(), &scrate);
+
+	let res: TokenStream = quote!(
+		#scrate_decl
+
+		#[derive(Clone, Copy, PartialEq, Eq)]
+		#[cfg_attr(feature = "std", derive(Debug))]
+		pub struct #name;
+		impl #scrate::sr_primitives::traits::GetNodeBlockType for #name {
+			type NodeBlock = #node_block;
+		}
+		impl #scrate::sr_primitives::traits::GetRuntimeBlockType for #name {
+			type RuntimeBlock = #block;
+		}
+
+		#outer_event
+
+		#outer_origin
+
+		#all_modules
+
+		#dispatch
+
+		#metadata
+
+		#outer_config
+
+		#inherent
+
+		#validate_unsigned
+	)
+	.into();
+
+	res
+}
+
+fn decl_validate_unsigned<'a>(
+	runtime: &'a Ident,
+	module_declarations: impl Iterator<Item = &'a ModuleDeclaration>,
+	scrate: &'a TokenStream2,
+) -> TokenStream2 {
+	let modules_tokens = module_declarations
+		.filter(|module_declaration| {
+			module_declaration
+				.module_parts()
+				.into_iter()
+				.any(|part| part.name == "ValidateUnsigned")
+		})
+		.map(|module_declaration| &module_declaration.name);
+	quote!(
+		#scrate::impl_outer_validate_unsigned!(
+			impl ValidateUnsigned for #runtime {
+				#( #modules_tokens )*
+			}
+		);
+	)
+}
+
+fn decl_outer_inherent<'a>(
+	block: &'a syn::TypePath,
+	unchecked_extrinsic: &'a syn::TypePath,
+	module_declarations: impl Iterator<Item = &'a ModuleDeclaration>,
+	scrate: &'a TokenStream2,
+) -> TokenStream2 {
+	let modules_tokens = module_declarations.filter_map(|module_declaration| {
+		let maybe_config_part = module_declaration
+			.module_parts()
+			.into_iter()
+			.filter(|part| part.name == "Inherent")
```

Comment:
This filtering is done almost in every method and leads to a lot of code duplication, please do this in a function.